### PR TITLE
[ROCm] Adding Cuda Alias to Gpu functions

### PR DESCRIPTION
**Pull request by jerryyin**
This PR is a follow-up to PR #24293. This PR depends on & includes code changes from PR #28116. Code updates for this PR are in the `tensorflow/core/util` path

Through utility functions defined in gpu_cuda_alias.h, this PR provides a Cuda* alias to existing Gpu* functions. The reason for this PR is to create backward compatibility to both the TensorFlow master branch as well as ROCm's develop-upstream branch. It allows TensorFlow master branch to slowly deprecate Cuda* aliased function, to the final goal of using GPU as the prefix to unify both Cuda and ROCm builds. All aliasing is done through perfect forwarding macro so that it is easy to maintain/take out.

Tests performed: Both Cuda/ROCm build successfully in both master and develop-upstream branch.

@whchung @deven-amd

### CODE REVIEW

**Comment by angersson** [comment]
I don't have the expertise to review this; removing myself.

**Comment by jerryyin** [comment]
Working on the broken builds...

File: tensorflow/core/util/gpu_launch_config.h [hunk 1]
```diff
@@ -308,18 +309,19 @@ inline Cuda3DLaunchConfig GetCuda3DLaunchConfig(
       block_size_limit);
   CHECK_EQ(err, cudaSuccess);
 #elif TENSORFLOW_USE_ROCM
-  // ROCM TODO re-enable this after hipOccupancyMaxPotentialBlockSize is
+  // XXX FIXME re-enable this after hipOccupancyMaxPotentialBlockSize is
```
Comment (whchung) [inline tensorflow/core/util/gpu_launch_config.h]: @jerryyin nit: XXX -> ROCM

**Review by whchung — CHANGES_REQUESTED** [review]

File: tensorflow/core/util/gpu_launch_config.h [hunk 2]
```diff
@@ -409,10 +415,10 @@ Status CudaLaunchKernel(void (*function)(Ts...), dim3 grid_dim, dim3 block_dim,
   }
   return Status::OK();
 }
-#endif  // GOOGLE_CUDA
+
```
Comment (whchung) [inline tensorflow/core/util/gpu_launch_config.h]: nit: extra whiteline

File: tensorflow/core/util/gpu_kernel_helper.h [hunk 3]
```diff
@@ -18,61 +18,70 @@ limitations under the License.
...
 #endif
 
 #if GOOGLE_CUDA
-#define GPU_LAUNCH_KERNEL(kernel, block_count, threads_per_block, shared_mem, \
-                          stream, ...)                                        \
-  TF_CHECK_OK(CudaLaunchKernel(kernel, block_count, threads_per_block,        \
-                               shared_mem, stream, __VA_ARGS__));
+#define GPU_LAUNCH_KERNEL(kernel, block_count, threads_per_block, \
+                          shared_mem, stream, ...) \
+  kernel<<<block_count, threads_per_block, shared_mem, stream>>>(__VA_ARGS__);
```
Comment (whchung) [inline tensorflow/core/util/gpu_kernel_helper.h]: @jerryyin per discussion we'll make `GPU_LAUNCH_KERNEL` macro obsolete and introduce `GpuLaunchKernel`, so maybe we can ignore adding more changes to this macro in this PR.

File: tensorflow/core/util/gpu_launch_config.h [hunk 4]
```diff
@@ -361,7 +365,9 @@ inline const cudaStream_t& GetCudaStream(OpKernelContext* context) {
                                                 ->GpuStreamMemberHack()));
   return *ptr;
 }
-#endif  // GOOGLE_CUDA
+#endif // GOOGLE_CUDA
+
+#define GetGpuStream(context) context->eigen_gpu_device().stream()
```
Comment (whchung) [inline tensorflow/core/util/gpu_launch_config.h]: @jerryyin this line doesn't really need to be moved, right?

**Comment by jerryyin** [comment]
Will post back to address review feedback after PR[439](https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/pull/439) approved. Will also integrate macro only change from GPU_LAUNCH_KERNEL to GpuLaunchKernel

File: tensorflow/core/util/gpu_cuda_alias.h [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+namespace tensorflow {
+
+#ifndef TENSORFLOW_USE_ROCM
```
[the remark this answers is not in the document]
Comment (jerryyin) [reply tensorflow/core/util/gpu_cuda_alias.h]: Will do

File: tensorflow/core/util/gpu_kernel_helper.h [hunk 6]
```diff
@@ -18,61 +18,70 @@ limitations under the License.
...
 #if GOOGLE_CUDA || TENSORFLOW_USE_ROCM
 
-#if GOOGLE_CUDA
-#include "cuda/include/cuda_fp16.h"
-#endif
+#include "tensorflow/core/util/gpu_cuda_alias.h"
 #include "tensorflow/core/util/gpu_device_functions.h"
 #include "tensorflow/core/util/gpu_launch_config.h"
 
+#if GPU_VERSION >= 7050
```
[the remark this answers is not in the document]
Comment (jerryyin) [reply tensorflow/core/util/gpu_kernel_helper.h]: My best guess is that the ROCm branch diverges from the master branch at certain point. This should be equivalent to original file's line 21 - 23. Will revert this back to original

File: tensorflow/core/util/gpu_kernel_helper.h [hunk 7]
```diff
@@ -18,61 +18,70 @@ limitations under the License.
...
 #endif
 
-// Deprecated, use 'for(int i : CudaGridRangeX(n))' instead.
+// Deprecated, use 'for(int i : GpuGridRangeX(n))' instead.
+#define GPU_1D_KERNEL_LOOP(i, n) \
+  for (int i : ::tensorflow::GpuGridRangeX<int>(n))
 #define CUDA_1D_KERNEL_LOOP(i, n) \
-  for (int i : ::tensorflow::CudaGridRangeX<int>(n))
-// Deprecated, use 'for(int i : CudaGridRange?(n))' instead.
+  for (int i : ::tensorflow::GpuGridRangeX<int>(n))
```
[the remark this answers is not in the document]
Comment (jerryyin) [reply tensorflow/core/util/gpu_kernel_helper.h]: Are you talking about macro CUDA_1D_KERNEL_LOOP? This macro is forwarded in line 49-50 with a GPU prefix.

File: tensorflow/core/util/gpu_kernel_helper.h [hunk 8]
```diff
@@ -18,61 +18,70 @@ limitations under the License.
...
+// Deprecated, use 'for(int i : GpuGridRange?(n))' instead.
+#define GPU_AXIS_KERNEL_LOOP(i, n, axis) \
+  for (int i : ::tensorflow::GpuGridRange##axis<int>(n))
 #define CUDA_AXIS_KERNEL_LOOP(i, n, axis) \
-  for (int i : ::tensorflow::CudaGridRange##axis<int>(n))
+  for (int i : ::tensorflow::GpuGridRange##axis<int>(n))
 
 #if GOOGLE_CUDA
 #define gpuSuccess cudaSuccess
+#define GPU_GET_ERROR_STRING(error) cudaGetErrorString(error)
```
[the remark this answers is not in the document]
Comment (jerryyin) [reply tensorflow/core/util/gpu_kernel_helper.h]: I think so, will address in next commit.

File: tensorflow/core/util/gpu_launch_config.h [hunk 9]
```diff
@@ -73,60 +74,62 @@ __global__ void MyKernel3D(Cuda3DLaunchConfig config, other_args...) {
...
+  Gpu3DLaunchConfig cfg6 = GetGpu3DLaunchConfig(4096, 4096, 100, d,
                                                   MyKernel1D, 0, 0);
   MyKernel3D <<<config.block_count,
                 config.thread_per_block, 0, d.stream()>>> (cfg6, other_args...);
 }
 
 // See the test for this for more example:
 //
-https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/util/gpu_kernel_helper_test.cu.cc
+https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/util/cuda_kernel_helper_test.cu.cc
```
[the remark this answers is not in the document]
Comment (jerryyin) [reply tensorflow/core/util/gpu_launch_config.h]: Another case where ROCm branch diverges. Will revert to original diff

File: tensorflow/core/util/gpu_launch_config.h [hunk 10]
```diff
@@ -308,18 +309,19 @@ inline Cuda3DLaunchConfig GetCuda3DLaunchConfig(
       block_size_limit);
   CHECK_EQ(err, cudaSuccess);
 #elif TENSORFLOW_USE_ROCM
-  // ROCM TODO re-enable this after hipOccupancyMaxPotentialBlockSize is
+  // XXX FIXME re-enable this after hipOccupancyMaxPotentialBlockSize is
   // implemented
-  // hipError_t err = hipOccupancyMaxPotentialBlockSize(
+  //hipError_t err = hipOccupancyMaxPotentialBlockSize(
```
[the remark this answers is not in the document]
Comment (jerryyin) [reply tensorflow/core/util/gpu_launch_config.h]: Will revert line 314 and 316 back to add the empty space

**Review by whchung — APPROVED** [review]

File: tensorflow/core/util/gpu_device_functions.h [hunk 11]
```diff
@@ -263,34 +284,34 @@ __device__ T CudaShuffleSync(unsigned mask, T value, int src_lane,
...
+  asm volatile("mov.b64 {%0,%1}, %2;" : "=r"(lo), "=r"(hi) : "d"(value));
+  hi = GpuShuffleSync(mask, hi, src_lane, width);
+  lo = GpuShuffleSync(mask, lo, src_lane, width);
+  asm volatile("mov.b64 %0, {%1,%2};" : "=d"(value) : "r"(lo), "r"(hi));
+  return value;
 #elif TENSORFLOW_USE_ROCM
-  auto tmp = static_cast<uint64_t>(value);
-  auto lo = static_cast<unsigned>(tmp);
-  auto hi = static_cast<unsigned>(tmp >> 32);
+  uint64_t tmp = static_cast<uint64_t>(value);
```
Comment (whchung) [inline tensorflow/core/util/gpu_device_functions.h]: @jerryyin to keep the PR simple, could you not change logic for these shuffle routines?

**Review by whchung — DISMISSED** [review]

File: tensorflow/core/util/gpu_device_functions.h (hunk 11 above)
Comment (jerryyin) [reply tensorflow/core/util/gpu_device_functions.h]: Thanks for reminding. The file version appeared here is before the weekly upstream sync, making it out of date. Checking out the current develop-upstream version gpu_device_functions.h seems to fix it.

**Comment by whchung** [comment]
@jerryyin multiple definition of `gpuGetErrorString(cudaError)'

**Comment by jerryyin** [comment]
> @jerryyin multiple definition of `gpuGetErrorString(cudaError)'

I was trying to forwarding the full set of `__host__` and `__device__` identifiers here according to Cuda's documentation [here](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__ERROR.html#group__CUDART__ERROR_1g4bc9e35a618dfd0877c29c8ee45148f1), but forgot that this is a header file with the function definition. When multiple compilation unit pulls it in, the compiler complains. I added `inline` identifier and local compilation works.

**Comment by chsigg** [comment]
Hi Zhuoran. I started the rename from our side, sorry that I didn't look more carefully at your approach first. Generally, please do not mix several changes into the same PR. It makes it hard to understand and review.
Could you please make a PR that introduces the macros only, and a follow-up to generate the various alias functions/macros/types, maybe even split into multiple PRs. At that point I can start preparing changes to use the new 'gpu' symbols.

**Comment by jerryyin** [comment]
> Hi Zhuoran. I started the rename from our side, sorry that I didn't look more carefully at your approach first. Generally, please do not mix several changes into the same PR. It makes it hard to understand and review.
> Could you please make a PR that introduces the macros only, and a follow-up to generate the various alias functions/macros/types, maybe even split into multiple PRs. At that point I can start preparing changes to use the new 'gpu' symbols.

@chsigg No worries, good feedback on breaking the PR. I'm planning on doing the following:

- [ ] One PR for create forwarding type macro, and corresponding changes in using the marco
- [ ] One PR for create forwarding host function macro, and corresponding changes in using the macro
- [ ] One PR for create forwarding device function macro, and corresponding changes in using the macro
- [ ] One PR for create GpuLaunchKernel
- [ ] One PR for updating GPU_LAUNCH_KERNEL to GpuLaunchKernel
- [ ] One PR for Removing GPU_LAUNCH_KERNEL

I will post back to here I as make progress, with the final goal of closing the large PR.

**Comment by jerryyin** [comment]
Closing the review as all sub-PRs are merged

**Pull request closed without merging**